Kernel/Arch/aarch64/PCI/Controller/BCM2712HostController.cpp

```diff
@@ -0,0 +1,263 @@
+    // CPU limit [31:20] @ [31:20]
+    // CPU base [31:20] @ [15:4]
+    u32 cpu_window_base_low_and_cpu_window_limit_low; // 0x4070
+
+    u8 unknown4[0x4080 - (0x4070 + 4)];
+
+    // CPU base [39:32] @ [7:0]
+    u32 cpu_window_base_high; // 0x4080
+
+    // CPU limit [39:32] @ [7:0]
+    u32 cpu_window_limit_high; // 0x4084
+
+    u8 unknown5[0x8000 - (0x4084 + 4)];
+
+    u8 config_space_window[0x1000];  // 0x8000
+    u32 config_space_window_address; // 0x9000
+
+    u8 unknown6[0x9310 - (0x9000 + 4)]; // 0x9000
+};
+static_assert(AssertSize<Registers, 0x9310>());
+
+class BCM2712HostController : public HostController {
+public:
+    static ErrorOr<NonnullOwnPtr<BCM2712HostController>> create(DeviceTree::Device const& device);
+
+private:
+    ErrorOr<VirtualAddress> get_config_window(BusNumber, DeviceNumber, FunctionNumber);
+
+    virtual void write8_field_locked(BusNumber, DeviceNumber, FunctionNumber, u32 field, u8 value) override;
+    virtual void write16_field_locked(BusNumber, DeviceNumber, FunctionNumber, u32 field, u16 value) override;
+    virtual void write32_field_locked(BusNumber, DeviceNumber, FunctionNumber, u32 field, u32 value) override;
+
+    virtual u8 read8_field_locked(BusNumber, DeviceNumber, FunctionNumber, u32 field) override;
+    virtual u16 read16_field_locked(BusNumber, DeviceNumber, FunctionNumber, u32 field) override;
+    virtual u32 read32_field_locked(BusNumber, DeviceNumber, FunctionNumber, u32 field) override;
+
+    explicit BCM2712HostController(PCI::Domain const&, Memory::TypedMapping<Registers volatile>);
+
+    Memory::TypedMapping<Registers volatile> m_registers;
+};
+
+ErrorOr<NonnullOwnPtr<BCM2712HostController>> BCM2712HostController::create(DeviceTree::Device const& device)
+{
+    auto domain = TRY(determine_pci_domain_for_devicetree_node(device.node(), device.node_name()));
+    auto registers_resource = TRY(device.get_resource(0));
+
+    if (registers_resource.size < sizeof(Registers))
+        return ERANGE;
+
+    auto registers = TRY(Memory::map_typed_writable<Registers volatile>(registers_resource.paddr));
+
+    // Deassert PERST# to initialize the link.
+    registers->control |= Registers::CONTROL_PERST_N;
+
+    microseconds_delay(100'000);
+
+    // Check the link state.
+    if ((registers->state & Registers::STATE_LINK_UP) != Registers::STATE_LINK_UP) {
+        dbgln("{}: Link down", device.node_name());
+
+        // We failed to initialize the link; assert PERST# again.
+        registers->control &= ~Registers::CONTROL_PERST_N;
+        return EIO;
+    }
+
+    dbgln("{}: Link up", device.node_name());
+
+    auto const* parent_node = device.node().parent();
+    if (parent_node == nullptr)
+        return EINVAL;
+
+    auto ranges = TRY(device.node().ranges());
+    for (size_t i = 0; i < ranges.entry_count(); i++) {
+        auto range = MUST(ranges.entry(i));
```

Comment:
Don't we have a nice `enumerate` helper?
(In AK)
or am I missing something about what that does

Reply:
We can't use `enumerate` here because the `ranges` function doesn't return something iterable. It instead lazily evaluates the entries when calling `entry(i)`.